README
# -*- coding: iso-latin-1-unix -*-
dbskkd-cdb readme/installation note in English
for version dbskkd-cdb-2.00
by Kenji Rikitake <kenji.rikitake@acm.org>
last updated 6-FEB-2009

* Summary

dbskkd-cdb is a dictionary server for SKK Japanese Input Method system.
For building the executable file, TinyCDB is required.

TinyCDB
http://www.corpit.ru/mjt/tinycdb.html

See skk-server-protocol.txt for the details.

This program is tested under FreeBSD 7.1-RELEASE.

* Installation procedure

1. makeskkcdb.sh converts a plaintext SKK dictionary to the CDB format.
   The default source file name is /usr/local/share/skk/SKK-JISYO.L
   and the converted cdb file will be placed in the current directory
   as SKK-JISYO.L.cdb.  This .cdb file should be placed with the filename
   specified in JISYO_FILE macro in dbskkd-cdb.c.

2. do 'make' to compile.  This require cdb.h and libcdb.a of TinyCDB.

3. do 'make install' to copy the compile executable image to the
   executable image directory.  The default directory is /usr/local/libexec.

* User privilege on installation

Assign a dedicated non-privileged user for this program and invoke with
the uid.

DO NOT invoke with the root or superuser privilege.

An example of inetd.conf is available at example-inetd.txt.
An example of daemontools run file is available at example-daemontools.txt.

* Bug reports and comments

The source code repository and issue tracker is available in Google Code at
http://code.google.com/p/dbskkd-cdb/

Inquiries and comments are welcome in Japanese and English.

* Acknowledgment

Daniel J. Bernstein for his cdb, Michael Tokarev for his TinyCDB.

SKK developers and contributors including:

Makahito Sato, Mikio Nakajima, Yukiyoshi Kameyama, Masaaki Sato,
Takashi Sakamoto, Fumitoshi Ukai, Toshinori Maeno, Yasuhiro Honda,
Hayao Nakahara, Takahiro Kikuchi, Hideo Matsumoto, Hidetaka Koie,
Takao Kawamura, Hideto Kihara, Rikishi Inakazu, Makoto Matsushita,
Atsushi Yamagata, Tatsuya Kinoshita, Ryosuke Nakai.

[end of README]



CHANGES
20161003
    LICENSE change to MIT
20131226
	Document changes only, no code change
	READMEJP revised and now READMEJP.md
	README deleted (read README.md instead)
20090206
	version: 2.00.
	READMEJP encoding changed to euc-jp
	skk-server-protocol.txt revised
	README added as a brief English document
20090131
	skk-server-protocol.txt added
	READMEJP revised and simplified
20090129
	Made CHANGES in reverse-chlonogical order
20090127
	makeskkcdb.sh now uses tinycdb's cdb
	instead of djb's cdbmake
20090126
	Removed ambiguity from LICENSE
	Attached LICENSE content in dbskkd-cdb.c
20090125
	inetd does not handle stderr well,
	so explicit error messages before crashing
	are all commented out
	(the fast-fail principle applied)
	setting VERBOSE_MSG as a compile flag
	will reenable the messages
	(still useful when used with ucspi-tcp)
	More comments on the code inserted
	DUMMYHOSTNAME made shorter
	NOTE: dictionary name is changed to JISYO_FILE
	      (previously JISHO -> now *JISYO*)
	      to match the SKK-JISYO.* filenames
	20090125-2 dbskkd-cdb.c performs macro replacement
		diesys(string) -> exit(1)
20090124
	dropped supporting server request '3'
	with actual hostname or IP address
	(the current syntax will be broken anyway
	 if IPv6 should be supported)
	since NO CLIENT uses this information for
	authentication or anything meaningful
	except for showing the information to the user
	when optionally requested, so the answer to
	the request is replaced by a fixed string
	defined in the DUMMYHOSTNAME macro
20090118
	version: 1.99-expr.
		strictly EXPERIMENTAL, NOT a release
	removed cdb-0.75 dependencies
	now tinycdb-0.76 required
	removed chroot, setuid, setgid calls from the program
	(more like 1.01)
	Japanese docs changed to reflect the status
20090114
	patches from Tatsuya Kinoshita (Debian package maintainer)
	* awk-locale.patch: Set LC_ALL to C for awk in skktocdbm.sh.
	* datasize.patch: Increase DATASIZE from 2048 to 4096 so that the
	  longest line of SKK-JISYO.L can be used.
	also Tatsuya pointed out the uint32 problem (which has been fixed)
	  of variable datalen
20090113
	BSD make dependencies resolved as pointed out in the following tweet:
		http://twitter.com/konosuke/status/1094925767
	cleanupcdb.sh added
	gcc-4.2.1 signedness warning cleared
		by quitting using unsigned chars
20000325
	Japanese README complete
	version: dbskkd-cdb-1.70dev.
20000312
	Changes from 1.01 to 1.70dev:
	* chroot()s to the SERVER_DIR
	* search() included into the main()'s command loop
	* reports pid to stderr
	* now uses cdb-0.75 source files instead of cdb-0.55's


dbskkd-cdb.c
/* 
 * dbskkd-cdb.c
 * SKK dictionary Server
 * with Daniel J. Bernstein's cdb database
 *
 * Copyright (c) 1998-2016 by Kenji Rikitake. All rights reserved.
 * See LICENSE of the distribution kit for the redistribution conditions.
 * The copy of the LICENSE file is attached at the end of this code.
 */

#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/param.h>
#include <sys/stat.h>
#include <sys/types.h>

/* tinycdb required */
#include "cdb.h"

/* 
 * dbskkd dictionary file 
 * default value: FreeBSD port japanese/ddskk
 * note: the macro was JISHO_FILE in 1.x
 *       spelling changed from JISHO to JISYO
 */
#ifndef JISYO_FILE
#define JISYO_FILE	"/usr/local/share/skk/SKK-JISYO.L.cdb"
#endif /* JISYO_FILE */

#define	BUFSIZE		(1024)	/* max size of a request */
#define DATASIZE	(4096)	/* max size of a result */

#define CLIENT_END	'0'
#define CLIENT_REQUEST	'1'
#define CLIENT_VERSION	'2'
#define CLIENT_HOST	'3'

#define SERVER_FOUND	'1'
#define SERVER_NOT_FOUND '4'

#define STDIN (fileno(stdin))
#define STDOUT (fileno(stdout))

/* these strings must be terminated with space */
#define VERSION "dbskkd-cdb-2.00 " 
#define DUMMYHOSTNAME "novalue: "

/*
 * define VERBOSE_MSG 
 * if explicit error messages are needed
 */
/* #undef VERBOSE_MSG */

/*
 * diesys() is replaced by exit(1)
 * when VERBOSE_MSG is not defined
 */
#ifdef VERBOSE_MSG
/* 
 * diesys() originally from DJB's cdb-0.55
 * modified by Kenji Rikitake
 */
void diesys(char *why)
{ 
  fprintf(stderr, "dbskkd-cdb: pid %d ", getpid());
  perror(why); 
  exit(1); 
}
#else /* !VERBOSE_MSG */
#define diesys(dummy_string) exit(1)
#endif /* VERBOSE_MSG */

/* main program */

int main(int argc, char *argv[])
{
  char combuf[BUFSIZE], data[DATASIZE];
  char *pbuf, *key, *p;
  struct cdb diccdb;
  int dicfd, ex, length;
  unsigned int keylen, datalen;

  /* open dictionary cdb file */
  if ((dicfd = open(JISYO_FILE, O_RDONLY, S_IRUSR)) < 0) {
    diesys("cannot open() the dictionary file " JISYO_FILE);
    /* NOTREACHED */
  }
  cdb_init(&diccdb, dicfd);

  /* command loop */
  /* set ex to non-zero for exiting */
  ex = 0;
  while (!ex) {

    length = read(STDIN, &combuf[0], BUFSIZE - 1);

    if (length < 0)
      diesys("read error from stdin");
    else if (length == 0) {
#ifdef VERBOSE_MSG
      fprintf(stderr, "dbskkd-cdb: pid %d end of file detected\n", getpid());
#endif /* VERBOSE_MSG */
      ex = 1;
      break;
    }

    /* parse request code */
    switch (combuf[0]) {

    case CLIENT_END:
#ifdef VERBOSE_MSG
      fprintf(stderr, 
	      "dbskkd-cdb: pid %d end of conversion requested\n", getpid());
#endif /* VERBOSE_MSG */
      ex = 1;
      break;

    case CLIENT_VERSION:
      if (write(STDOUT, VERSION, sizeof(VERSION) - 1) < 0)
        diesys("write error (version string)"); 
      break;

    case CLIENT_HOST:
      if (write(STDOUT, DUMMYHOSTNAME, sizeof(DUMMYHOSTNAME) - 1) < 0)
          diesys("write error (dummyhostname string)"); 
      break;

    case CLIENT_REQUEST:
      /* get size of key */
      key = &combuf[1];
      for (pbuf = &combuf[1];
           *pbuf != ' ' && pbuf != &combuf[length - 1];
           pbuf++) {}
      keylen = pbuf - &combuf[1];
      if (keylen <= 0)
        diesys("invalid keysize");

      /* lookup the cdb database */
      switch (cdb_find(&diccdb, key, keylen)) {

      case -1: /* error */
        diesys("fatal error on cdb_find()");

      case 1: /* found */
        if ((datalen = cdb_datalen(&diccdb)) >= DATASIZE - 2)
  	  diesys("searched entry too long");

	/* generate the answer string */
	p = data;
	*p++ = SERVER_FOUND;
        if (cdb_read(&diccdb, p, datalen, cdb_datapos(&diccdb)) < 0)
  	  diesys("read error (seeked dictionary data)");
        p += datalen;
        *p = '\n';

        /* sending found code and the result data string with LF */
        if (write(STDOUT, data, datalen + 2) < 0)
  	  diesys("write error (converted data)");
        break;

      case 0: /* NOT found */
        /* generate the answer string */
        combuf[0] = SERVER_NOT_FOUND;
        *pbuf = '\n';

        /* sending error code and the key string with LF */
        /* this action is required by skkinput */
        if (write(STDOUT, combuf, keylen + 2) < 0)
  	  diesys("write error (NOT_FOUND message)");
        break;

      default: /* unknown cdb return value */ 
        diesys("unknown return value from cdb_seek()");
      }
      break;
      /* end CLIENT_REQUEST switch statement*/

    /* unknown request code */
    default: 
#ifdef VERBOSE_MSG
      fprintf(stderr, 
	      "dbskkd-cdb: pid %d unknown client request code = %d\n",
	       getpid(), combuf[0]);
#endif /* VERBOSE_MSG */
      exit(1);
    } /* end request code parsing */
  }

  /* close dictionary db file */
  cdb_free(&diccdb);
  if (close(dicfd) != 0)
    diesys("cannot close() the dictionary file");

  exit(0);
}

/* end of program */

/* 
 License statement for dbskkd-cdb

 The MIT License (MIT)
 Copyright © 2016 Kenji Rikitake <kenji.rikitake@acm.org>

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the “Software”), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 */

/*
 For more information on the dbskkd-cdb project, please refer to 
 https://github.com/jj1bdx/dbskkd-cdb/
 */
